File: src/lib/kinds.ts
// Routed-kind ↔ URL-segment mapping. Single TS source of truth.
//
// Pure TypeScript (no `astro:content` import) so `astro.config.ts` can import
// it. Python mirror: `pancratius/kinds.py`; cross-language guard:
// `audit/python/kind_segments.py`.

/** The routable content kinds, in canonical order. */
export const ROUTED_KINDS = ["book", "poem", "project", "video", "message"] as const;

/** A single routable content kind. */
export type RoutedKind = typeof ROUTED_KINDS[number];

/** The corpus work kinds. Projects and videos route but aren't works. */
export const CORPUS_WORK_KINDS = ["book", "poem"] as const;

/** A corpus work kind: convertible, downloadable, and pairable by language. */
export type CorpusWorkKind = typeof CORPUS_WORK_KINDS[number];

/** The URL segments a routed kind can map to. */
export type RoutedSegment = "books" | "poetry" | "projects" | "videos" | "messages";

/** Routed kind → URL segment. Every segment is a structural English noun; the
 * displayed section label is localized separately («Послания» / Epistles for
 * messages). */
export const SEGMENT_OF: Record<RoutedKind, RoutedSegment> = {
  book:    "books",
  poem:    "poetry",
  project: "projects",
  video:   "videos",
  message: "messages",
};

/** URL segment → routed kind (inverse of `SEGMENT_OF`). */
export const KIND_OF_SEGMENT: Record<RoutedSegment, RoutedKind> = {
  books:    "book",
  poetry:   "poem",
  projects: "project",
  videos:   "video",
  messages: "message",
};

/** True when an arbitrary path segment is one of the routed content segments. */
export function isRoutedSegment(value: string): value is RoutedSegment {
  return Object.prototype.hasOwnProperty.call(KIND_OF_SEGMENT, value);
}

File: pancratius/kinds.py
"""Canonical routed-kind -> URL-segment mapping (Python side).

This mirrors ``src/lib/kinds.ts``. Python cannot import the TS module and the
config can't import Python, so the mapping necessarily exists once per language.
``audit/python/kind_segments.py`` is the cross-language guard: it asserts this
dict equals the ``SEGMENT_OF`` map in ``src/lib/kinds.ts``.
"""

from __future__ import annotations

from typing import Literal

type RoutedKind = Literal["book", "poem", "project", "video", "message"]
type CorpusWorkKind = Literal["book", "poem"]
type RoutedSegment = Literal["books", "poetry", "projects", "videos", "messages"]

# Routed content kind -> structural-noun URL segment. Includes `project` (themed
# sections under /projects/) and `video` (catalogued YouTube/other-platform
# videos under /videos/) and `message` (dated posts under /messages/, labelled
# «Послания» / Epistles in the UI) — all route and appear in the sitemap.
# Routing breadth != convertible-work scope. Every segment is an English noun;
# the displayed section label is localized separately.
SEGMENT_OF: dict[RoutedKind, RoutedSegment] = {
    "book": "books",
    "poem": "poetry",
    "project": "projects",
    "video": "videos",
    "message": "messages",
}

# The kinds that are convertible/downloadable corpus works — the source of truth
# for "which kinds the import/converter pipeline handles and which kinds get a
# download matrix". Projects and videos route but are not works (no DOCX-import,
# no PDF/EPUB matrix), so they are intentionally NOT here. This tuple is a
# subset of SEGMENT_OF's keys.
CORPUS_WORK_KINDS: tuple[CorpusWorkKind, ...] = ("book", "poem")

# URL segment -> routed kind (inverse of SEGMENT_OF).
KIND_OF_SEGMENT: dict[RoutedSegment, RoutedKind] = {
    segment: kind for kind, segment in SEGMENT_OF.items()
}
